### PR TITLE
Quote fee adjustment

### PR DESCRIPTION
**Pull request by squadgazzz**
While testing https://github.com/cowprotocol/services/issues/2642, @fhenneke noticed a discrepancy in the quote fee amount between the new API and Auction JSON, where the former looks more correct since it doesn't use `f64` calculation: https://github.com/cowprotocol/services/blob/quote-fee-adjustment/crates/orderbook/src/database/fee_policies.rs#L91-L103

The same calculation approach is now used in both places.

### CODE REVIEW

File: crates/autopilot/src/infra/persistence/dto/quote.rs [hunk 1]
```diff
@@ -1,25 +1,46 @@
...
     })
 }
 
 #[derive(Debug, thiserror::Error)]
-#[error("amount overflow")]
-pub struct AmountOverflow;
+pub enum QuoteError {
+    #[error("amount overflow")]
+    AmountOverflow,
+    #[error("invalid input")]
```
Comment (Mateo-mro) [inline crates/autopilot/src/infra/persistence/dto/quote.rs:40]: Can you be more precise, mentioning that it cannot be converted to a valid rational number

File: crates/autopilot/src/infra/persistence/dto/quote.rs [hunk 2]
```diff
@@ -1,25 +1,46 @@
...
+            .ok_or(QuoteError::AmountOverflow)?
+            .checked_div(&sell_token_price)
+            .ok_or(QuoteError::DivisionByZero)?,
+    )
+    .map_err(QuoteError::Error)?;
     Ok(domain::Quote {
         order_uid: domain::OrderUid(quote.order_uid.0),
         sell_amount: big_decimal_to_u256(&quote.sell_amount)
-            .ok_or(AmountOverflow)?
+            .ok_or(QuoteError::AmountOverflow)?
```
Comment (Mateo-mro) [inline crates/autopilot/src/infra/persistence/dto/quote.rs:27]: I would separate this error into two:
- `AmountOverflow` for operation
- `AmountOverflow` for transforming the result into an `U256`
WDYT?

File: crates/autopilot/src/infra/persistence/dto/quote.rs [hunk 3]
```diff
@@ -1,25 +1,46 @@
...
-pub struct AmountOverflow;
+pub enum QuoteError {
+    #[error("amount overflow")]
+    AmountOverflow,
+    #[error("invalid input")]
+    InvalidInput,
+    #[error("division by zero")]
+    DivisionByZero,
+    #[error(transparent)]
+    Error(#[from] anyhow::Error),
```
Comment (Mateo-mro) [inline crates/autopilot/src/infra/persistence/dto/quote.rs:45]: for what is this one needed? :thinking:

**Review by Mateo-mro — APPROVED** [review]
Yes, please :pray: doing operations directly with `f64` is not a good idea.

File: crates/autopilot/src/infra/persistence/dto/quote.rs (hunk 2 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/dto/quote.rs]: Makes sense. Added another error.

File: crates/autopilot/src/infra/persistence/dto/quote.rs (hunk 1 above)
Comment (squadgazzz) [reply crates/autopilot/src/infra/persistence/dto/quote.rs]: Done

**Comment by fhenneke** [comment]
I am in favor of moving to the correct rounding here.

Note that this PR will require a change to the circuit breaker. The current version of the circuit breaker has to duplicate the inaccurate quote fee computation to replicate scores from the driver. With this change, I need to change the computation of quote fees to the accurate version.

This will overall make the computation of scores more consistent, see #2797.

**Review by fleupold — APPROVED** [review]

**Comment by harisang** [comment]
I still don't understand though why we want to force solvers to compute themselves the quoted buy/sell amount (and potentially running into rounding issues) instead of having the autopilot report these numbers explicitly (and doing the operations internally whatever way we want).

**Comment by fleupold** [comment]
You mean concretely why not replace


``
quote: {
 buy:
 sell:
 fee:
}
```

with
```
quote: {
  buy:
  sell:
}
```

where fee is already incorporated? @sunce86 you were the one designing the quote object. Do you remember why you chose to expose the fee separately?

**Comment by sunce86** [comment]
No strong reason. At that time I wasn't sure how exactly the Quote object will be used in the driver and whether we will need fee explicitly. Also found it cleaner from the API perspective to understand the quote if fee exists separately.

**Pull request merged**